File: backend/models/backtest.py
"""Backtest models — strategy templates, backtest configurations, and run results."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone

from sqlalchemy import Float, Index, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


class StrategyTemplate(Base):
    __tablename__ = "strategy_templates"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(String(36), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    category: Mapped[str] = mapped_column(String(50), nullable=False)  # e.g. "momentum", "mean_reversion", "custom"
    strategy_code: Mapped[str] = mapped_column(Text, nullable=False)  # python snippet for backtrader strategy
    params_json: Mapped[str | None] = mapped_column(Text, nullable=True)  # JSON: {"sma_fast": 10, "sma_slow": 30}
    created_at: Mapped[str] = mapped_column(Text, default=_now)
    updated_at: Mapped[str] = mapped_column(Text, default=_now, onupdate=_now)

    __table_args__ = (
        Index("idx_strategy_user", "user_id"),
        Index("idx_strategy_category", "category"),
    )


class BacktestRun(Base):
    __tablename__ = "backtest_runs"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(String(36), nullable=False)
    strategy_template_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")  # pending, running, completed, failed
    symbol: Mapped[str] = mapped_column(String(20), nullable=False)  # e.g. "AAPL", "SPY"
    timeframe: Mapped[str] = mapped_column(String(10), nullable=False, default="1d")  # 1d, 1h, 15m
    start_date: Mapped[str] = mapped_column(String(30), nullable=False)
    end_date: Mapped[str] = mapped_column(String(30), nullable=False)
    initial_cash: Mapped[float] = mapped_column(Float, nullable=False, default=10000.0)
    commission: Mapped[float] = mapped_column(Float, nullable=False, default=0.001)  # 0.1%

    # Results
    total_return_pct: Mapped[float | None] = mapped_column(Float, nullable=True)
    sharpe_ratio: Mapped[float | None] = mapped_column(Float, nullable=True)
    max_drawdown_pct: Mapped[float | None] = mapped_column(Float, nullable=True)
    win_rate_pct: Mapped[float | None] = mapped_column(Float, nullable=True)
    total_trades: Mapped[int | None] = mapped_column(Integer, nullable=True)
    final_value: Mapped[float | None] = mapped_column(Float, nullable=True)

    # Raw results (JSON snapshots for replay)
    equity_curve_json: Mapped[str | None] = mapped_column(Text, nullable=True)  # [{date, value}, ...]
    trades_json: Mapped[str | None] = mapped_column(Text, nullable=True)  # [{date, action, price, size}, ...]
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[str] = mapped_column(Text, default=_now)
    started_at: Mapped[str | None] = mapped_column(Text, nullable=True)
    completed_at: Mapped[str | None] = mapped_column(Text, nullable=True)

    __table_args__ = (
        Index("idx_backtest_user", "user_id"),
        Index("idx_backtest_status", "status"),
        Index("idx_backtest_user_status", "user_id", "status"),
    )


class PaperTrade(Base):
    __tablename__ = "paper_trades"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(String(36), nullable=False)
    backtest_run_id: Mapped[str | None] = mapped_column(String(36), nullable=True)  # optional, if spawned from backtest
    strategy_template_id: Mapped[str | None] = mapped_column(String(36), nullable=True)
    symbol: Mapped[str] = mapped_column(String(20), nullable=False)
    action: Mapped[str] = mapped_column(String(10), nullable=False)  # buy, sell
    quantity: Mapped[float] = mapped_column(Float, nullable=False)
    price: Mapped[float] = mapped_column(Float, nullable=False)
    order_type: Mapped[str] = mapped_column(String(20), nullable=False, default="market")  # market, limit
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="filled")  # pending, filled, cancelled
    pnl: Mapped[float | None] = mapped_column(Float, nullable=True)
    created_at: Mapped[str] = mapped_column(Text, default=_now)
    updated_at: Mapped[str] = mapped_column(Text, default=_now, onupdate=_now)

    __table_args__ = (
        Index("idx_paper_user", "user_id"),
        Index("idx_paper_user_symbol", "user_id", "symbol"),
    )

File: frontend/src/api/backtest.ts
import { api } from "./client";

export interface StrategyTemplate {
  id: string;
  user_id: string;
  name: string;
  description: string | null;
  category: string;
  strategy_code: string;
  params_json: string | null;
  created_at: string;
  updated_at: string;
}

export interface BacktestRun {
  id: string;
  user_id: string;
  strategy_template_id: string | null;
  status: "pending" | "running" | "completed" | "failed";
  symbol: string;
  timeframe: string;
  start_date: string;
  end_date: string;
  initial_cash: number;
  commission: number;
  total_return_pct: number | null;
  sharpe_ratio: number | null;
  max_drawdown_pct: number | null;
  win_rate_pct: number | null;
  total_trades: number | null;
  final_value: number | null;
  equity_curve_json: string | null;
  trades_json: string | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface PaperTrade {
  id: string;
  user_id: string;
  backtest_run_id: string | null;
  strategy_template_id: string | null;
  symbol: string;
  action: "buy" | "sell";
  quantity: number;
  price: number;
  order_type: string;
  status: string;
  pnl: number | null;
  created_at: string;
  updated_at: string;
}

export interface PaperPortfolio {
  positions: Array<{
    symbol: string;
    quantity: number;
    cost_basis: number;
    current_value: number;
    pnl: number;
  }>;
  total_pnl: number;
  position_count: number;
}

function qs(params: Record<string, string | number | undefined>): string {
  const p = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null) p.set(k, String(v));
  }
  const s = p.toString();
  return s ? `?${s}` : "";
}

export const backtestApi = {
  // Strategy templates
  listStrategies: (offset = 0, limit = 20) =>
    api<{ strategies: StrategyTemplate[]; count: number }>(`/backtest/strategies${qs({ offset, limit })}`),

  getStrategy: (id: string) =>
    api<StrategyTemplate>(`/backtest/strategies/${id}`),

  createStrategy: (data: {
    name: string;
    category: string;
    strategy_code: string;
    description?: string;
    params_json?: string;
  }) =>
    api<StrategyTemplate>("/backtest/strategies", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  updateStrategy: (id: string, data: Record<string, string | null>) =>
    api<StrategyTemplate>(`/backtest/strategies/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    }),

  deleteStrategy: (id: string) =>
    api<void>(`/backtest/strategies/${id}`, { method: "DELETE" }),

  // Backtest runs
  startRun: (data: {
    symbol: string;
    start_date: string;
    end_date: string;
    strategy_template_id?: string;
    initial_cash?: number;
    commission?: number;
    timeframe?: string;
  }) =>
    api<BacktestRun>("/backtest/runs", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  listRuns: (offset = 0, limit = 20) =>
    api<{ runs: BacktestRun[]; count: number }>(`/backtest/runs${qs({ offset, limit })}`),

  getRun: (id: string) =>
    api<BacktestRun>(`/backtest/runs/${id}`),

  // Paper trading
  createPaperTrade: (data: {
    symbol: string;
    action: "buy" | "sell";
    quantity: number;
    price: number;
    order_type?: string;
    backtest_run_id?: string;
    strategy_template_id?: string;
  }) =>
    api<PaperTrade>("/backtest/paper-trades", {
      method: "POST",
      body: JSON.stringify(data),
    }),

  listPaperTrades: (offset = 0, limit = 50) =>
    api<{ trades: PaperTrade[]; count: number }>(`/backtest/paper-trades${qs({ offset, limit })}`),

  paperPortfolio: () =>
    api<PaperPortfolio>("/backtest/paper-portfolio"),
};